plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,418 @@
+      final ORID bucketOrid = bucketEntityAdapter.decode(tx.getBucket().getEntityMetadata().getId());
+      sqlParams.put("bucket", bucketOrid);
+      worker.rebuild(tx);
+    }
+  }
+
+  /**
+   * Inner class (non static) that encapsulates the work, as metadata builder is stateful.
+   */
+  private class Worker
+  {
+    private final boolean update;
+
+    private final String sql;
+
+    private final Map<String, Object> sqlParams;
+
+    private final StorageFacet storageFacet;
+
+    private final MavenFacet mavenFacet;
+
+    private final MavenPathParser mavenPathParser;
+
+    private final MetadataBuilder metadataBuilder;
+
+    private final MetadataUpdater metadataUpdater;
+
+    private final DocumentBuilderFactory documentBuilderFactory;
+
+    public Worker(final Repository repository,
+                  final boolean update,
+                  final String sql,
+                  final Map<String, Object> sqlParams)
+    {
+      this.update = update;
+      this.sql = sql;
+      this.sqlParams = sqlParams;
+      this.storageFacet = repository.facet(StorageFacet.class);
+      this.mavenFacet = repository.facet(MavenFacet.class);
+      this.mavenPathParser = mavenFacet.getMavenPathParser();
+      this.metadataBuilder = new MetadataBuilder();
+      this.metadataUpdater = new MetadataUpdater(mavenFacet);
+      this.documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    }
+
+    public void rebuild(final StorageTx tx)
+    {
+      final ODatabaseDocumentTx readDatabase = tx.getDb();
+      readDatabase.command(
+          new OSQLAsynchQuery<ODocument>(
+              sql,
+              new OCommandResultListener()
+              {
+                String currentGroupId = null;
+
+                @Override
+                public boolean result(Object iRecord) {
+                  try {
+                    final ODocument doc = (ODocument) iRecord;
+                    final String groupId = doc.field("groupId", OType.STRING);
+                    final String artifactId = doc.field("artifactId", OType.STRING);
+                    final Set<String> baseVersions = doc.field("baseVersions", OType.EMBEDDEDSET);
+
+                    final boolean groupChange = !Objects.equals(currentGroupId, groupId);
+                    if (groupChange) {
+                      if (currentGroupId != null) {
+                        processMetadata(metadataMavenPath(currentGroupId, null, null), metadataBuilder.onExitGroupId());
+                      }
+                      currentGroupId = groupId;
+                      metadataBuilder.onEnterGroupId(groupId);
+                    }
+                    rebuildAbV((ORID) sqlParams.get("bucket"), groupId, artifactId, baseVersions);
+                    return true;
+                  }
+                  finally {
+                    ODatabaseRecordThreadLocal.INSTANCE.set(readDatabase); // always reset thread DB
```

Comment:
Might be time to discuss this along with https://github.com/sonatype/nexus-pro/blob/master/plugins/security/nexus-usertoken-plugin/src/main/java/com/sonatype/nexus/usertoken/plugin/store/orient/UserTokenRecordEntityAdapter.java#L91-91
This seems like the kind of detail that would be easy to forget when we start expanding our data access operations.